In [1]:
# data preprocessing
from datasets import load_dataset
from transformers import AutoTokenizer

# load IMDB dataset
def load_data_and_tokenize():
    # load IMDB dataset 
    dataset = load_dataset("imdb")
    train_data = dataset["train"]
    test_data = dataset["test"]

    train_data = train_data.shuffle(seed=42).select(range(2048))
    test_data = test_data.shuffle(seed=42).select(range(2048))

    # tokenize the data
    tokenizer = AutoTokenizer.from_pretrained(model_name)
    def tokenize_function(examples):
        return tokenizer(examples["text"], padding="max_length", truncation=True, max_length=max_len)
    tokenized_train_data = train_data.map(tokenize_function, batched=True)
    tokenized_test_data = test_data.map(tokenize_function, batched=True)


    # delete text column
    tokenized_train_data = tokenized_train_data.remove_columns(["text"])
    tokenized_test_data  = tokenized_test_data.remove_columns(["text"])

    # python list -> torch tensor
    tokenized_train_data.set_format(
        type="torch",
        columns=["input_ids", "attention_mask", "label"]
    )
    tokenized_test_data.set_format(
        type="torch",
        columns=["input_ids", "attention_mask", "label"]
    )

    return tokenized_train_data, tokenized_test_data

In [2]:
import torch

def validation_step(model, validation_loader, device):
    model.eval()
    total_loss = 0.0
    total_correct = 0
    total_count = 0

    with torch.no_grad():
        for batch in validation_loader:
            input_ids = batch["input_ids"].to(device)
            attention_mask = batch["attention_mask"].to(device)
            labels = batch["label"].to(device)  # shape: (batch,)

            outputs = model(
                input_ids=input_ids,
                attention_mask=attention_mask,
                labels=labels,
            )

            loss = outputs.loss               # scalar
            logits = outputs.logits           # shape: (batch, num_labels)

            total_loss += loss.item() * input_ids.size(0)
            preds = torch.argmax(logits, dim=-1)
            total_correct += (preds == labels).sum().item()
            total_count += input_ids.size(0)

    avg_loss = total_loss / total_count
    avg_acc = total_correct / total_count
    return  avg_loss, avg_acc

In [7]:
from xml.parsers.expat import model

import torch
import time
import wandb
from tqdm import tqdm


def train(model, train_dataset, validation_dataset, num_epochs, batch_size, learning_rate, num_workers, optimizer_name = "AdamW", compile_mode = False):

    device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
    model.to(device)

    if optimizer_name == "SGD":
        optimizer = torch.optim.SGD(model.parameters(), lr=learning_rate, momentum=0.9)
    elif optimizer_name == "Adam":
        optimizer = torch.optim.Adam(model.parameters(), lr=learning_rate)
    elif optimizer_name == "AdamW":
        optimizer = torch.optim.AdamW(model.parameters(), lr=learning_rate)
    else:
        raise ValueError(f"Unsupported optimizer: {optimizer_name}")
    
    train_loader = torch.utils.data.DataLoader(
        train_dataset, batch_size=batch_size, shuffle=True, num_workers=num_workers
    )
    validation_loader = torch.utils.data.DataLoader(
        validation_dataset, batch_size=batch_size, shuffle=False, num_workers=num_workers
    )

    if compile_mode:
        model = torch.compile(model) 

    initial_val_loss, initial_val_acc = validation_step(model, validation_loader, device)
    print(f'Initial validation loss: {initial_val_loss}, accuracy: {initial_val_acc}')
    # wandb.log({'val_loss': initial_val_loss, 'val_acc': initial_val_acc})

    for epoch in tqdm(list(range(num_epochs))):
        model.train()

        train_losses = 0.0
        total_correct = 0
        data_time = 0.0
        compute_time = 0.0

        end_time = time.time()

        # 5c) Training loop
        for batch in train_loader:
            ############################################################
            # STUDENT IMPLEMENTATION START
            # 1. clear the gradients stored in optimizer.
            # 2. compute the loss by calling compute_loss
            # 3. backpropagate the loss
            # 4. call optimizer to update the model parameters
            ############################################################

            # data time
            data_time += time.time() - end_time
            compute_start_time = time.time()
            
            # clear gradients
            optimizer.zero_grad()

            # compute loss
            input_ids = batch["input_ids"].to(device)
            attention_mask = batch["attention_mask"].to(device)
            labels = batch["label"].to(device)

            outputs = model(
                input_ids=input_ids,
                attention_mask=attention_mask,
                labels=labels,
            )

            loss = outputs.loss               # scalar
            logits = outputs.logits           # shape: (batch, num_labels)

            # backpropagate
            loss.backward()

            # update model parameters
            optimizer.step()
            torch.cuda.synchronize()


            # record training loss and accuracy
            end_time = time.time()
            compute_time += end_time - compute_start_time
            train_losses += loss.item() * input_ids.size(0)
            preds = torch.argmax(logits, dim=-1)
            total_correct += (preds == labels).sum().item()

        
        # validation step
        test_loss, test_acc = validation_step(model, validation_loader, device)
        avg_train_loss = train_losses / len(train_dataset)
        train_acc = total_correct / len(train_dataset)

        wandb.log({
            'train/loss': avg_train_loss,
            'train/acc': train_acc,
            'test/loss': test_loss,
            'test/acc': test_acc,
            'time/data_loading': data_time,
            'time/compute': compute_time,
            'time/epoch': data_time + compute_time
        })
        print(f'Epoch {epoch}: train loss: {avg_train_loss}, data loading time: {data_time:.2f}s, compute time: {compute_time:.2f}s')

    return

In [16]:
import torch
import time
import wandb
from tqdm import tqdm


def count_trainable_params(model: torch.nn.Module):
    trainable_tensors = 0
    trainable_elems = 0
    for p in model.parameters():
        if p.requires_grad:
            trainable_tensors += 1
            trainable_elems += p.numel()
    return trainable_tensors, trainable_elems


def count_params_with_grads(model: torch.nn.Module):
    grad_tensors = 0
    grad_elems = 0
    for p in model.parameters():
        if p.grad is not None:
            grad_tensors += 1
            grad_elems += p.grad.numel()
    return grad_tensors, grad_elems


def train_and_count_paramters(
    model,
    train_dataset,
    validation_dataset,
    num_epochs,
    batch_size,
    learning_rate,
    num_workers,
    optimizer_name="AdamW",
    compile_mode=False,
):
    device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
    model.to(device)

    # Optimizer
    if optimizer_name == "SGD":
        optimizer = torch.optim.SGD(model.parameters(), lr=learning_rate, momentum=0.9)
    elif optimizer_name == "Adam":
        optimizer = torch.optim.Adam(model.parameters(), lr=learning_rate)
    elif optimizer_name == "AdamW":
        optimizer = torch.optim.AdamW(model.parameters(), lr=learning_rate)
    else:
        raise ValueError(f"Unsupported optimizer: {optimizer_name}")

    train_loader = torch.utils.data.DataLoader(
        train_dataset, batch_size=batch_size, shuffle=True, num_workers=num_workers
    )
    validation_loader = torch.utils.data.DataLoader(
        validation_dataset, batch_size=batch_size, shuffle=False, num_workers=num_workers
    )

    if compile_mode:
        model = torch.compile(model)

    # ---- Count trainable params (once) ----
    trainable_tensors, trainable_elems = count_trainable_params(model)
    print(f"Trainable parameters: {trainable_elems:,} (across {trainable_tensors} tensors)")
    wandb.log({
        "model/trainable_param_tensors": trainable_tensors,
        "model/trainable_param_elems": trainable_elems,
    })

    # initial_val_loss, initial_val_acc = validation_step(model, validation_loader, device)
    # print(f"Initial validation loss: {initial_val_loss}, accuracy: {initial_val_acc}")

    for epoch in tqdm(list(range(num_epochs))):
        model.train()

        train_losses = 0.0
        total_correct = 0
        data_time = 0.0
        compute_time = 0.0

        end_time = time.time()

        # Training loop
        for batch_idx, batch in enumerate(train_loader):
            # data time
            data_time += time.time() - end_time
            compute_start_time = time.time()

            optimizer.zero_grad()

            input_ids = batch["input_ids"].to(device)
            attention_mask = batch["attention_mask"].to(device)
            labels = batch["label"].to(device)

            outputs = model(
                input_ids=input_ids,
                attention_mask=attention_mask,
                labels=labels,
            )

            loss = outputs.loss
            logits = outputs.logits

            loss.backward()

            # ---- Count params with grads (after epoch) ----
            if epoch == 0 and batch_idx == 0:
                grad_tensors, grad_elems = count_params_with_grads(model)
                print(f"Parameters with gradients (after one backward): "f"{grad_elems:,} (across {grad_tensors} tensors)")

            optimizer.step()

            if device.type == "cuda":
                torch.cuda.synchronize()

            # record training loss and accuracy
            end_time = time.time()
            compute_time += end_time - compute_start_time
            train_losses += loss.item() * input_ids.size(0)
            preds = torch.argmax(logits, dim=-1)
            total_correct += (preds == labels).sum().item()


        # validation step
        test_loss, test_acc = validation_step(model, validation_loader, device)
        avg_train_loss = train_losses / len(train_dataset)
        train_acc = total_correct / len(train_dataset)

        wandb.log({
            "train/loss": avg_train_loss,
            "train/acc": train_acc,
            "test/loss": test_loss,
            "test/acc": test_acc,
            "time/data_loading": data_time,
            "time/compute": compute_time,
            "time/epoch": data_time + compute_time,
            "model/grad_param_tensors": grad_tensors,
            "model/grad_param_elems": grad_elems,
        })

        print(
            f"Epoch {epoch}: train loss: {avg_train_loss}, "
            f"data loading time: {data_time:.2f}s, compute time: {compute_time:.2f}s, "
            f"params w/ grads: {grad_elems:,} (across {grad_tensors} tensors)"
        )

    return

In [3]:
from csv import writer
import os
import time
import torch
import wandb
from tqdm import tqdm
from torch.profiler import profile, ProfilerActivity, tensorboard_trace_handler
from torch.utils.tensorboard import SummaryWriter

def train_profiler(
    model, train_dataset, validation_dataset,
    num_epochs, batch_size, learning_rate, num_workers,
    profiler_dir="/content/drive/MyDrive/tb_profiler"  # <-- save to Google Drive
):
    # Mount Google Drive (Colab)
    try:
        from google.colab import drive
        if not os.path.exists("/content/drive"):
            os.makedirs("/content/drive", exist_ok=True)
        # If not mounted yet, this will prompt once
        if not os.path.exists("/content/drive/MyDrive"):
            drive.mount("/content/drive")
    except Exception as e:
        print(f"[Warning] Drive mount skipped/failed: {e}")
        return

    # Select device
    device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
    model.to(device)

    # Initialize optimizer
    optimizer = torch.optim.AdamW(model.parameters(), lr=learning_rate)

    # Create DataLoaders
    train_loader = torch.utils.data.DataLoader(
        train_dataset,
        batch_size=batch_size,
        shuffle=True,
        num_workers=num_workers,
        pin_memory=torch.cuda.is_available(),
        persistent_workers=(num_workers > 0),
    )

    validation_loader = torch.utils.data.DataLoader(
        validation_dataset,
        batch_size=batch_size,
        shuffle=False,
        num_workers=num_workers,
        pin_memory=torch.cuda.is_available(),
        persistent_workers=(num_workers > 0),
    )

    # Initial validation before training
    initial_val_loss, initial_val_acc = validation_step(model, validation_loader, device)
    print(f"Initial validation loss: {initial_val_loss}, accuracy: {initial_val_acc}")

    # Make sure the profiler directory exists on Drive
    os.makedirs(profiler_dir, exist_ok=True)

    for epoch in tqdm(range(num_epochs)):
        model.train()

        train_losses = 0.0
        total_correct = 0
        
        data_time = 0.0
        compute_time = 0.0

        end_time = time.time()

        # Save trace files under:
        # /content/drive/MyDrive/tb_profiler/workers_{num_workers}/epoch_{epoch}/
        trace_dir = os.path.join(profiler_dir, f"workers_{num_workers}", f"epoch_{epoch}")
        os.makedirs(trace_dir, exist_ok=True)
        
        # Profile the entire epoch
        with profile(
            schedule=torch.profiler.schedule(wait=1, warmup=1, active=3, repeat=1),
            on_trace_ready=tensorboard_trace_handler(trace_dir),
            record_shapes=True,
            with_stack=True,
        ) as prof:

            for step, batch in enumerate(train_loader):
                prof.step()
                # Measure data loading time
                data_time += time.time() - end_time
                compute_start_time = time.time()

                optimizer.zero_grad()

                # Move batch to device
                input_ids = batch["input_ids"].to(device, non_blocking=True)
                attention_mask = batch["attention_mask"].to(device, non_blocking=True)
                labels = batch["label"].to(device, non_blocking=True)

                # Forward pass
                outputs = model(
                    input_ids=input_ids,
                    attention_mask=attention_mask,
                    labels=labels,
                )

                loss = outputs.loss
                logits = outputs.logits

                # Backward pass and parameter update
                loss.backward()
                optimizer.step()

                # Synchronize to get more accurate compute timing
                if torch.cuda.is_available():
                    torch.cuda.synchronize()

                end_time = time.time()
                compute_time += end_time - compute_start_time

                # Track metrics
                train_losses += loss.item() * input_ids.size(0)
                preds = torch.argmax(logits, dim=-1)
                total_correct += (preds == labels).sum().item()

        # Validation step after each epoch
        test_loss, test_acc = validation_step(model, validation_loader, device)

        avg_train_loss = train_losses / len(train_dataset)
        train_acc = total_correct / len(train_dataset)

        wandb.log({
            "train/loss": avg_train_loss,
            "train/acc": train_acc,
            "test/loss": test_loss,
            "test/acc": test_acc,
            "time/data_loading": data_time,
            "time/compute": compute_time,
            "time/epoch": data_time + compute_time
        })

        print(
            f"Epoch {epoch}: train loss {avg_train_loss:.4f}, "
            f"data {data_time:.2f}s, compute {compute_time:.2f}s"
        )
        print(f"[Profiler] Trace saved to: {trace_dir}")

In [ ]:
#C1 & C2
import wandb
from transformers import AutoModelForSequenceClassification

model_name = "distilbert-base-uncased"
max_len = 256
batch_size = 32
num_epochs = 5
learning_rate = 1e-4
optimizer = "AdamW"
num_workers = 2
compile_mode = False


wandb.init(project="hpml-hw2-llm")
wandb.config.update({
    "model_name": model_name,
    "max_len": max_len,
    "batch_size": batch_size,
    "lr": learning_rate,
    "optimizer": optimizer,
    "num_workers": num_workers,
    "epochs": num_epochs,
    "compile_mode": compile_mode
})

model = AutoModelForSequenceClassification.from_pretrained(
    model_name,
    num_labels=2  
)

train_dataset, validation_dataset = load_data_and_tokenize()

train(model, train_dataset, validation_dataset)

Loading weights:   0%|          | 0/100 [00:00<?, ?it/s]

DistilBertForSequenceClassification LOAD REPORT from: distilbert-base-uncased
Key                     | Status     | 
------------------------+------------+-
vocab_projector.bias    | UNEXPECTED | 
vocab_transform.weight  | UNEXPECTED | 
vocab_layer_norm.bias   | UNEXPECTED | 
vocab_transform.bias    | UNEXPECTED | 
vocab_layer_norm.weight | UNEXPECTED | 
classifier.weight       | MISSING    | 
pre_classifier.bias     | MISSING    | 
pre_classifier.weight   | MISSING    | 
classifier.bias         | MISSING    | 

Notes:
- UNEXPECTED	:can be ignored when loading from different task/architecture; not ok if you expect identical arch.
- MISSING	:those params were newly initialized because missing from the checkpoint. Consider training on your downstream task.


Map:   0%|          | 0/2048 [00:00<?, ? examples/s]

Initial validation loss: 0.6935234414413571, accuracy: 0.49755859375


 20%|██        | 1/5 [00:59<03:58, 59.59s/it]

Epoch 0: train loss: 0.5130960145033896, data loading time: 0.17s, compute time: 44.77s


 40%|████      | 2/5 [01:58<02:57, 59.25s/it]

Epoch 1: train loss: 0.23558978090295568, data loading time: 0.18s, compute time: 43.91s


 60%|██████    | 3/5 [02:57<01:58, 59.12s/it]

Epoch 2: train loss: 0.10688533589564031, data loading time: 0.18s, compute time: 44.08s


 80%|████████  | 4/5 [03:56<00:59, 59.15s/it]

Epoch 3: train loss: 0.07262088460993255, data loading time: 0.17s, compute time: 44.21s


100%|██████████| 5/5 [04:56<00:00, 59.20s/it]

Epoch 4: train loss: 0.05211467490516952, data loading time: 0.18s, compute time: 44.27s


In [19]:
# C3
from transformers import AutoModelForSequenceClassification
import wandb

model_name = "distilbert-base-uncased"
max_len = 256
batch_size = 32
num_epochs = 5
learning_rate = 1e-4
optimizer_name = "AdamW"
compile_mode = False

worker_list = [0, 2, 4, 8]

for num_workers in worker_list:

    wandb.init(
        project="hpml-hw2-llm",
        name=f"workers_{num_workers}",
        reinit=True
    )

    wandb.config.update({
        "model_name": model_name,
        "max_len": max_len,
        "batch_size": batch_size,
        "lr": learning_rate,
        "optimizer": optimizer_name,
        "num_workers": num_workers,
        "epochs": num_epochs,
        "compile_mode": compile_mode
    })

    model = AutoModelForSequenceClassification.from_pretrained(
        model_name,
        num_labels=2
    )

    train_dataset, validation_dataset = load_data_and_tokenize()

    train(
        model,
        train_dataset,
        validation_dataset,
        num_epochs=num_epochs,
        batch_size=batch_size,
        learning_rate=learning_rate,
        num_workers=num_workers
    )

    wandb.finish()

Loading weights:   0%|          | 0/100 [00:00<?, ?it/s]

DistilBertForSequenceClassification LOAD REPORT from: distilbert-base-uncased
Key                     | Status     | 
------------------------+------------+-
vocab_projector.bias    | UNEXPECTED | 
vocab_transform.weight  | UNEXPECTED | 
vocab_layer_norm.bias   | UNEXPECTED | 
vocab_transform.bias    | UNEXPECTED | 
vocab_layer_norm.weight | UNEXPECTED | 
classifier.weight       | MISSING    | 
pre_classifier.bias     | MISSING    | 
pre_classifier.weight   | MISSING    | 
classifier.bias         | MISSING    | 

Notes:
- UNEXPECTED	:can be ignored when loading from different task/architecture; not ok if you expect identical arch.
- MISSING	:those params were newly initialized because missing from the checkpoint. Consider training on your downstream task.


Initial validation loss: 0.693833751603961, accuracy: 0.4931640625


 20%|██        | 1/5 [00:58<03:52, 58.05s/it]

Epoch 0: train loss: 0.5337188690900803, data loading time: 0.17s, compute time: 42.76s


 40%|████      | 2/5 [01:57<02:55, 58.66s/it]

Epoch 1: train loss: 0.2892174094449729, data loading time: 0.17s, compute time: 44.34s


 60%|██████    | 3/5 [02:56<01:57, 58.93s/it]

Epoch 2: train loss: 0.14064890882582404, data loading time: 0.17s, compute time: 44.30s


 80%|████████  | 4/5 [03:55<00:58, 58.92s/it]

Epoch 3: train loss: 0.061657004327571485, data loading time: 0.17s, compute time: 44.03s


100%|██████████| 5/5 [04:54<00:00, 58.91s/it]

Epoch 4: train loss: 0.07639277319867688, data loading time: 0.17s, compute time: 44.25s


test/acc,▆█▅▄▁
test/loss,▂▁▃▆█
time/compute,▁██▇█
time/data_loading,█▆▄▁▇
time/epoch,▁██▇█
train/acc,▁▆▇██
train/loss,█▄▂▁▁
test/acc,0.83691
test/loss,0.62229
time/compute,44.25371
time/data_loading,0.17232


Loading weights:   0%|          | 0/100 [00:00<?, ?it/s]

DistilBertForSequenceClassification LOAD REPORT from: distilbert-base-uncased
Key                     | Status     | 
------------------------+------------+-
vocab_projector.bias    | UNEXPECTED | 
vocab_transform.weight  | UNEXPECTED | 
vocab_layer_norm.bias   | UNEXPECTED | 
vocab_transform.bias    | UNEXPECTED | 
vocab_layer_norm.weight | UNEXPECTED | 
classifier.weight       | MISSING    | 
pre_classifier.bias     | MISSING    | 
pre_classifier.weight   | MISSING    | 
classifier.bias         | MISSING    | 

Notes:
- UNEXPECTED	:can be ignored when loading from different task/architecture; not ok if you expect identical arch.
- MISSING	:those params were newly initialized because missing from the checkpoint. Consider training on your downstream task.


Initial validation loss: 0.6989230597391725, accuracy: 0.49658203125


 20%|██        | 1/5 [00:59<03:58, 59.73s/it]

Epoch 0: train loss: 0.4564724110532552, data loading time: 0.25s, compute time: 44.78s


 40%|████      | 2/5 [01:59<02:58, 59.57s/it]

Epoch 1: train loss: 0.2285492083756253, data loading time: 0.29s, compute time: 44.21s


 60%|██████    | 3/5 [02:58<01:58, 59.42s/it]

Epoch 2: train loss: 0.09994586918037385, data loading time: 0.26s, compute time: 44.07s


 80%|████████  | 4/5 [03:57<00:59, 59.40s/it]

Epoch 3: train loss: 0.07465747684909729, data loading time: 0.26s, compute time: 44.17s


100%|██████████| 5/5 [04:57<00:00, 59.44s/it]

Epoch 4: train loss: 0.049401460666558705, data loading time: 0.25s, compute time: 44.28s


test/acc,▄█▁▄▆
test/loss,▁▂██▅
time/compute,█▂▁▂▃
time/data_loading,▁█▃▃▁
time/epoch,█▃▁▂▃
train/acc,▁▅▇██
train/loss,█▄▂▁▁
test/acc,0.85547
test/loss,0.44031
time/compute,44.27803
time/data_loading,0.25046


Loading weights:   0%|          | 0/100 [00:00<?, ?it/s]

DistilBertForSequenceClassification LOAD REPORT from: distilbert-base-uncased
Key                     | Status     | 
------------------------+------------+-
vocab_projector.bias    | UNEXPECTED | 
vocab_transform.weight  | UNEXPECTED | 
vocab_layer_norm.bias   | UNEXPECTED | 
vocab_transform.bias    | UNEXPECTED | 
vocab_layer_norm.weight | UNEXPECTED | 
classifier.weight       | MISSING    | 
pre_classifier.bias     | MISSING    | 
pre_classifier.weight   | MISSING    | 
classifier.bias         | MISSING    | 

Notes:
- UNEXPECTED	:can be ignored when loading from different task/architecture; not ok if you expect identical arch.
- MISSING	:those params were newly initialized because missing from the checkpoint. Consider training on your downstream task.
/usr/local/lib/python3.12/dist-packages/torch/utils/data/dataloader.py:424: UserWarning: This DataLoader will create 4 worker processes in total. Our suggested max number of worker in current system is 2, which is smaller than what th

Initial validation loss: 0.6938119484111667, accuracy: 0.494140625


 20%|██        | 1/5 [01:00<04:00, 60.05s/it]

Epoch 0: train loss: 0.49538419442251325, data loading time: 0.31s, compute time: 44.96s


 40%|████      | 2/5 [01:59<02:59, 59.80s/it]

Epoch 1: train loss: 0.26158950570970774, data loading time: 0.40s, compute time: 44.13s


 60%|██████    | 3/5 [02:59<01:59, 59.65s/it]

Epoch 2: train loss: 0.1200634556444129, data loading time: 0.33s, compute time: 44.09s


 80%|████████  | 4/5 [03:58<00:59, 59.71s/it]

Epoch 3: train loss: 0.07242856285665766, data loading time: 0.39s, compute time: 44.30s


100%|██████████| 5/5 [04:58<00:00, 59.72s/it]

Epoch 4: train loss: 0.04315870517530129, data loading time: 0.31s, compute time: 44.26s


test/acc,▇▁▅█▆
test/loss,▁▃▃▆█
time/compute,█▁▁▃▂
time/data_loading,▁█▃█▁
time/epoch,█▂▁▃▂
train/acc,▁▅▇██
train/loss,█▄▂▁▁
test/acc,0.85693
test/loss,0.52042
time/compute,44.26073
time/data_loading,0.30976


Loading weights:   0%|          | 0/100 [00:00<?, ?it/s]

DistilBertForSequenceClassification LOAD REPORT from: distilbert-base-uncased
Key                     | Status     | 
------------------------+------------+-
vocab_projector.bias    | UNEXPECTED | 
vocab_transform.weight  | UNEXPECTED | 
vocab_layer_norm.bias   | UNEXPECTED | 
vocab_transform.bias    | UNEXPECTED | 
vocab_layer_norm.weight | UNEXPECTED | 
classifier.weight       | MISSING    | 
pre_classifier.bias     | MISSING    | 
pre_classifier.weight   | MISSING    | 
classifier.bias         | MISSING    | 

Notes:
- UNEXPECTED	:can be ignored when loading from different task/architecture; not ok if you expect identical arch.
- MISSING	:those params were newly initialized because missing from the checkpoint. Consider training on your downstream task.
/usr/local/lib/python3.12/dist-packages/torch/utils/data/dataloader.py:424: UserWarning: This DataLoader will create 8 worker processes in total. Our suggested max number of worker in current system is 2, which is smaller than what th

Initial validation loss: 0.7033390533179045, accuracy: 0.50341796875


 20%|██        | 1/5 [01:00<04:01, 60.31s/it]

Epoch 0: train loss: 0.4811117588542402, data loading time: 0.50s, compute time: 44.92s


 40%|████      | 2/5 [02:00<03:00, 60.23s/it]

Epoch 1: train loss: 0.20743203564779833, data loading time: 0.43s, compute time: 44.25s


 60%|██████    | 3/5 [03:00<01:59, 59.99s/it]

Epoch 2: train loss: 0.09542822661751416, data loading time: 0.43s, compute time: 44.05s


 80%|████████  | 4/5 [04:00<01:00, 60.01s/it]

Epoch 3: train loss: 0.07108737679300248, data loading time: 0.44s, compute time: 44.29s


100%|██████████| 5/5 [05:00<00:00, 60.05s/it]

Epoch 4: train loss: 0.036062204595509684, data loading time: 0.45s, compute time: 44.25s


test/acc,█▁█▇█
test/loss,▁▆▄▇█
time/compute,█▃▁▃▃
time/data_loading,█▁▁▂▃
time/epoch,█▃▁▃▃
train/acc,▁▆▇▇█
train/loss,█▄▂▂▁
test/acc,0.8667
test/loss,0.49415
time/compute,44.2549
time/data_loading,0.45051


In [27]:
!ls /content/drive/MyDrive/

In [20]:
# C5: sweep batch size and learning rate (grid)
from transformers import AutoModelForSequenceClassification
import wandb
import torch

model_name = "distilbert-base-uncased"
max_len = 256
num_epochs = 5
optimizer_name = "AdamW"
compile_mode = False
num_workers = 2  

batch_size_list = [16, 32, 64]
lr_list = [5e-5, 1e-4, 5e-4]

for batch_size in batch_size_list:
    for learning_rate in lr_list:

        run_name = f"bs{batch_size}_lr{learning_rate:g}"

        wandb.init(
            project="hpml-hw2-llm",
            group="C5_bs_lr",
            name=run_name,
            reinit=True
        )

        wandb.config.update({
            "exp": "C5",
            "model_name": model_name,
            "max_len": max_len,
            "batch_size": batch_size,
            "lr": learning_rate,
            "optimizer": optimizer_name,
            "num_workers": num_workers,
            "epochs": num_epochs,
            "compile_mode": compile_mode
        })

        # (optional) reduce cross-run GPU memory fragmentation
        if torch.cuda.is_available():
            torch.cuda.empty_cache()

        model = AutoModelForSequenceClassification.from_pretrained(
            model_name,
            num_labels=2
        )

        train_dataset, validation_dataset = load_data_and_tokenize()

        train(
            model,
            train_dataset,
            validation_dataset,
            num_epochs=num_epochs,
            batch_size=batch_size,
            learning_rate=learning_rate,
            num_workers=num_workers
        )

        wandb.finish()

Loading weights:   0%|          | 0/100 [00:00<?, ?it/s]

DistilBertForSequenceClassification LOAD REPORT from: distilbert-base-uncased
Key                     | Status     | 
------------------------+------------+-
vocab_projector.bias    | UNEXPECTED | 
vocab_transform.weight  | UNEXPECTED | 
vocab_layer_norm.bias   | UNEXPECTED | 
vocab_transform.bias    | UNEXPECTED | 
vocab_layer_norm.weight | UNEXPECTED | 
classifier.weight       | MISSING    | 
pre_classifier.bias     | MISSING    | 
pre_classifier.weight   | MISSING    | 
classifier.bias         | MISSING    | 

Notes:
- UNEXPECTED	:can be ignored when loading from different task/architecture; not ok if you expect identical arch.
- MISSING	:those params were newly initialized because missing from the checkpoint. Consider training on your downstream task.


Initial validation loss: 0.6941632693633437, accuracy: 0.498046875


 20%|██        | 1/5 [00:59<03:58, 59.72s/it]

Epoch 0: train loss: 0.4265438426518813, data loading time: 0.39s, compute time: 44.77s


 40%|████      | 2/5 [01:59<02:58, 59.61s/it]

Epoch 1: train loss: 0.2192980075487867, data loading time: 0.43s, compute time: 44.96s


 60%|██████    | 3/5 [02:59<01:59, 59.72s/it]

Epoch 2: train loss: 0.11382919530296931, data loading time: 0.39s, compute time: 45.31s


 80%|████████  | 4/5 [03:58<00:59, 59.67s/it]

Epoch 3: train loss: 0.09831138894878677, data loading time: 0.40s, compute time: 45.06s


100%|██████████| 5/5 [04:58<00:00, 59.71s/it]

Epoch 4: train loss: 0.03904780888296955, data loading time: 0.40s, compute time: 45.22s


test/acc,█▇▅▄▁
test/loss,▁▁▃▃█
time/compute,▁▃█▅▇
time/data_loading,▁█▁▃▃
time/epoch,▁▄█▅▇
train/acc,▁▅▇▇█
train/loss,█▄▂▂▁
test/acc,0.84326
test/loss,0.70736
time/compute,45.21509
time/data_loading,0.40056


Loading weights:   0%|          | 0/100 [00:00<?, ?it/s]

DistilBertForSequenceClassification LOAD REPORT from: distilbert-base-uncased
Key                     | Status     | 
------------------------+------------+-
vocab_projector.bias    | UNEXPECTED | 
vocab_transform.weight  | UNEXPECTED | 
vocab_layer_norm.bias   | UNEXPECTED | 
vocab_transform.bias    | UNEXPECTED | 
vocab_layer_norm.weight | UNEXPECTED | 
classifier.weight       | MISSING    | 
pre_classifier.bias     | MISSING    | 
pre_classifier.weight   | MISSING    | 
classifier.bias         | MISSING    | 

Notes:
- UNEXPECTED	:can be ignored when loading from different task/architecture; not ok if you expect identical arch.
- MISSING	:those params were newly initialized because missing from the checkpoint. Consider training on your downstream task.


Initial validation loss: 0.6911128181964159, accuracy: 0.53759765625


 20%|██        | 1/5 [01:00<04:00, 60.15s/it]

Epoch 0: train loss: 0.4764860343420878, data loading time: 0.39s, compute time: 45.76s


 40%|████      | 2/5 [01:59<02:59, 59.92s/it]

Epoch 1: train loss: 0.2530503803427564, data loading time: 0.43s, compute time: 45.05s


 60%|██████    | 3/5 [02:59<01:59, 59.82s/it]

Epoch 2: train loss: 0.11624152150398004, data loading time: 0.40s, compute time: 45.06s


 80%|████████  | 4/5 [03:59<00:59, 59.84s/it]

Epoch 3: train loss: 0.08288642728621198, data loading time: 0.43s, compute time: 45.21s


100%|██████████| 5/5 [04:59<00:00, 59.86s/it]

Epoch 4: train loss: 0.07652519591829332, data loading time: 0.41s, compute time: 45.24s


test/acc,▅▇█▁▃
test/loss,▂▁▂▆█
time/compute,█▁▁▃▃
time/data_loading,▁█▄█▆
time/epoch,█▁▁▃▃
train/acc,▁▆███
train/loss,█▄▂▁▁
test/acc,0.84668
test/loss,0.55901
time/compute,45.24021
time/data_loading,0.414


Loading weights:   0%|          | 0/100 [00:00<?, ?it/s]

DistilBertForSequenceClassification LOAD REPORT from: distilbert-base-uncased
Key                     | Status     | 
------------------------+------------+-
vocab_projector.bias    | UNEXPECTED | 
vocab_transform.weight  | UNEXPECTED | 
vocab_layer_norm.bias   | UNEXPECTED | 
vocab_transform.bias    | UNEXPECTED | 
vocab_layer_norm.weight | UNEXPECTED | 
classifier.weight       | MISSING    | 
pre_classifier.bias     | MISSING    | 
pre_classifier.weight   | MISSING    | 
classifier.bias         | MISSING    | 

Notes:
- UNEXPECTED	:can be ignored when loading from different task/architecture; not ok if you expect identical arch.
- MISSING	:those params were newly initialized because missing from the checkpoint. Consider training on your downstream task.


Initial validation loss: 0.6971090300939977, accuracy: 0.501953125


 20%|██        | 1/5 [00:59<03:56, 59.23s/it]

Epoch 0: train loss: 0.7057520891539752, data loading time: 0.39s, compute time: 45.39s


 40%|████      | 2/5 [01:57<02:56, 58.95s/it]

Epoch 1: train loss: 0.6959057752974331, data loading time: 0.38s, compute time: 44.73s


 60%|██████    | 3/5 [02:56<01:57, 58.86s/it]

Epoch 2: train loss: 0.6970053645782173, data loading time: 0.39s, compute time: 44.73s


 80%|████████  | 4/5 [03:55<00:58, 58.87s/it]

Epoch 3: train loss: 0.6940148314461112, data loading time: 0.43s, compute time: 44.80s


100%|██████████| 5/5 [04:54<00:00, 58.91s/it]

Epoch 4: train loss: 0.693677878472954, data loading time: 0.42s, compute time: 44.82s


test/acc,█▁█▁▁
test/loss,▁█▁▃▇
time/compute,█▁▁▂▂
time/data_loading,▁▁▂█▆
time/epoch,█▁▁▂▂
train/acc,▁█▂▃▄
train/loss,█▂▃▁▁
test/acc,0.49658
test/loss,0.69876
time/compute,44.81715
time/data_loading,0.42381


Loading weights:   0%|          | 0/100 [00:00<?, ?it/s]

DistilBertForSequenceClassification LOAD REPORT from: distilbert-base-uncased
Key                     | Status     | 
------------------------+------------+-
vocab_projector.bias    | UNEXPECTED | 
vocab_transform.weight  | UNEXPECTED | 
vocab_layer_norm.bias   | UNEXPECTED | 
vocab_transform.bias    | UNEXPECTED | 
vocab_layer_norm.weight | UNEXPECTED | 
classifier.weight       | MISSING    | 
pre_classifier.bias     | MISSING    | 
pre_classifier.weight   | MISSING    | 
classifier.bias         | MISSING    | 

Notes:
- UNEXPECTED	:can be ignored when loading from different task/architecture; not ok if you expect identical arch.
- MISSING	:those params were newly initialized because missing from the checkpoint. Consider training on your downstream task.


Initial validation loss: 0.6933295521885157, accuracy: 0.49365234375


 20%|██        | 1/5 [00:59<03:58, 59.71s/it]

Epoch 0: train loss: 0.46413463866338134, data loading time: 0.24s, compute time: 44.82s


 40%|████      | 2/5 [01:59<02:58, 59.49s/it]

Epoch 1: train loss: 0.2343594161211513, data loading time: 0.25s, compute time: 44.07s


 60%|██████    | 3/5 [02:58<01:58, 59.38s/it]

Epoch 2: train loss: 0.11766291994717903, data loading time: 0.27s, compute time: 44.09s


 80%|████████  | 4/5 [03:57<00:59, 59.45s/it]

Epoch 3: train loss: 0.05839868877956178, data loading time: 0.25s, compute time: 44.26s


100%|██████████| 5/5 [04:57<00:00, 59.44s/it]

Epoch 4: train loss: 0.03506365529392497, data loading time: 0.29s, compute time: 44.22s


test/acc,▃▁█▅▅
test/loss,▂▄▁▅█
time/compute,█▁▁▃▂
time/data_loading,▁▁▄▂█
time/epoch,█▁▁▃▃
train/acc,▁▆▇██
train/loss,█▄▂▁▁
test/acc,0.8584
test/loss,0.51341
time/compute,44.21523
time/data_loading,0.28967


Loading weights:   0%|          | 0/100 [00:00<?, ?it/s]

DistilBertForSequenceClassification LOAD REPORT from: distilbert-base-uncased
Key                     | Status     | 
------------------------+------------+-
vocab_projector.bias    | UNEXPECTED | 
vocab_transform.weight  | UNEXPECTED | 
vocab_layer_norm.bias   | UNEXPECTED | 
vocab_transform.bias    | UNEXPECTED | 
vocab_layer_norm.weight | UNEXPECTED | 
classifier.weight       | MISSING    | 
pre_classifier.bias     | MISSING    | 
pre_classifier.weight   | MISSING    | 
classifier.bias         | MISSING    | 

Notes:
- UNEXPECTED	:can be ignored when loading from different task/architecture; not ok if you expect identical arch.
- MISSING	:those params were newly initialized because missing from the checkpoint. Consider training on your downstream task.


Initial validation loss: 0.7032604785636067, accuracy: 0.50341796875


 20%|██        | 1/5 [00:59<03:59, 59.92s/it]

Epoch 0: train loss: 0.43578679906204343, data loading time: 0.25s, compute time: 45.01s


 40%|████      | 2/5 [01:59<02:59, 59.69s/it]

Epoch 1: train loss: 0.2359445936162956, data loading time: 0.29s, compute time: 44.18s


 60%|██████    | 3/5 [02:58<01:59, 59.51s/it]

Epoch 2: train loss: 0.10243978284415789, data loading time: 0.24s, compute time: 44.04s


 80%|████████  | 4/5 [03:58<00:59, 59.53s/it]

Epoch 3: train loss: 0.0748754489395651, data loading time: 0.24s, compute time: 44.30s


100%|██████████| 5/5 [04:57<00:00, 59.57s/it]

Epoch 4: train loss: 0.03987217872236215, data loading time: 0.25s, compute time: 44.29s


test/acc,▁█▄▄▃
test/loss,▂▁▃█▆
time/compute,█▂▁▃▃
time/data_loading,▂█▁▁▂
time/epoch,█▂▁▃▃
train/acc,▁▅▇██
train/loss,█▄▂▂▁
test/acc,0.84912
test/loss,0.46994
time/compute,44.28747
time/data_loading,0.2506


Loading weights:   0%|          | 0/100 [00:00<?, ?it/s]

DistilBertForSequenceClassification LOAD REPORT from: distilbert-base-uncased
Key                     | Status     | 
------------------------+------------+-
vocab_projector.bias    | UNEXPECTED | 
vocab_transform.weight  | UNEXPECTED | 
vocab_layer_norm.bias   | UNEXPECTED | 
vocab_transform.bias    | UNEXPECTED | 
vocab_layer_norm.weight | UNEXPECTED | 
classifier.weight       | MISSING    | 
pre_classifier.bias     | MISSING    | 
pre_classifier.weight   | MISSING    | 
classifier.bias         | MISSING    | 

Notes:
- UNEXPECTED	:can be ignored when loading from different task/architecture; not ok if you expect identical arch.
- MISSING	:those params were newly initialized because missing from the checkpoint. Consider training on your downstream task.


Initial validation loss: 0.6992904609069228, accuracy: 0.50244140625


 20%|██        | 1/5 [00:59<03:57, 59.31s/it]

Epoch 0: train loss: 0.705184823833406, data loading time: 0.24s, compute time: 44.66s


 40%|████      | 2/5 [01:57<02:55, 58.63s/it]

Epoch 1: train loss: 0.7120898766443133, data loading time: 0.25s, compute time: 43.63s


 60%|██████    | 3/5 [02:55<01:56, 58.31s/it]

Epoch 2: train loss: 0.7007704088464379, data loading time: 0.24s, compute time: 43.69s


 80%|████████  | 4/5 [03:53<00:58, 58.23s/it]

Epoch 3: train loss: 0.6996846459805965, data loading time: 0.25s, compute time: 43.79s


100%|██████████| 5/5 [04:51<00:00, 58.33s/it]

Epoch 4: train loss: 0.6956442110240459, data loading time: 0.28s, compute time: 43.87s


test/acc,▁▁███
test/loss,▄▅█▁▂
time/compute,█▁▁▂▃
time/data_loading,▁▃▁▃█
time/epoch,█▁▁▂▃
train/acc,▂█▁▆▇
train/loss,▅█▃▃▁
test/acc,0.50342
test/loss,0.69442
time/compute,43.87112
time/data_loading,0.27807


Loading weights:   0%|          | 0/100 [00:00<?, ?it/s]

DistilBertForSequenceClassification LOAD REPORT from: distilbert-base-uncased
Key                     | Status     | 
------------------------+------------+-
vocab_projector.bias    | UNEXPECTED | 
vocab_transform.weight  | UNEXPECTED | 
vocab_layer_norm.bias   | UNEXPECTED | 
vocab_transform.bias    | UNEXPECTED | 
vocab_layer_norm.weight | UNEXPECTED | 
classifier.weight       | MISSING    | 
pre_classifier.bias     | MISSING    | 
pre_classifier.weight   | MISSING    | 
classifier.bias         | MISSING    | 

Notes:
- UNEXPECTED	:can be ignored when loading from different task/architecture; not ok if you expect identical arch.
- MISSING	:those params were newly initialized because missing from the checkpoint. Consider training on your downstream task.


Initial validation loss: 0.6934567000716925, accuracy: 0.50048828125


 20%|██        | 1/5 [00:59<03:58, 59.66s/it]

Epoch 0: train loss: 0.45635085785761476, data loading time: 0.16s, compute time: 44.58s


 40%|████      | 2/5 [01:58<02:58, 59.38s/it]

Epoch 1: train loss: 0.24480140977539122, data loading time: 0.21s, compute time: 43.98s


 60%|██████    | 3/5 [02:58<01:58, 59.37s/it]

Epoch 2: train loss: 0.12758319964632392, data loading time: 0.16s, compute time: 44.15s


 80%|████████  | 4/5 [03:57<00:59, 59.27s/it]

Epoch 3: train loss: 0.09952137889922597, data loading time: 0.16s, compute time: 43.92s


100%|██████████| 5/5 [04:56<00:00, 59.34s/it]

Epoch 4: train loss: 0.05664514549425803, data loading time: 0.17s, compute time: 44.10s


test/acc,█▄▁▇▇
test/loss,▁▁▅▅█
time/compute,█▂▃▁▃
time/data_loading,▁█▁▂▂
time/epoch,█▂▃▁▃
train/acc,▁▅▇▇█
train/loss,█▄▂▂▁
test/acc,0.87842
test/loss,0.41702
time/compute,44.10165
time/data_loading,0.16594


Loading weights:   0%|          | 0/100 [00:00<?, ?it/s]

DistilBertForSequenceClassification LOAD REPORT from: distilbert-base-uncased
Key                     | Status     | 
------------------------+------------+-
vocab_projector.bias    | UNEXPECTED | 
vocab_transform.weight  | UNEXPECTED | 
vocab_layer_norm.bias   | UNEXPECTED | 
vocab_transform.bias    | UNEXPECTED | 
vocab_layer_norm.weight | UNEXPECTED | 
classifier.weight       | MISSING    | 
pre_classifier.bias     | MISSING    | 
pre_classifier.weight   | MISSING    | 
classifier.bias         | MISSING    | 

Notes:
- UNEXPECTED	:can be ignored when loading from different task/architecture; not ok if you expect identical arch.
- MISSING	:those params were newly initialized because missing from the checkpoint. Consider training on your downstream task.


Initial validation loss: 0.6909114345908165, accuracy: 0.50634765625


 20%|██        | 1/5 [00:59<03:58, 59.66s/it]

Epoch 0: train loss: 0.4797264481894672, data loading time: 0.20s, compute time: 44.69s


 40%|████      | 2/5 [01:58<02:58, 59.42s/it]

Epoch 1: train loss: 0.24519358505494893, data loading time: 0.18s, compute time: 43.90s


 60%|██████    | 3/5 [02:58<01:58, 59.27s/it]

Epoch 2: train loss: 0.13402460829820484, data loading time: 0.17s, compute time: 43.91s


 80%|████████  | 4/5 [03:57<00:59, 59.33s/it]

Epoch 3: train loss: 0.08036680991062894, data loading time: 0.17s, compute time: 44.17s


100%|██████████| 5/5 [04:56<00:00, 59.32s/it]

Epoch 4: train loss: 0.05743850795261096, data loading time: 0.16s, compute time: 44.01s


test/acc,▂▁▄█▆
test/loss,▁▁▂▆█
time/compute,█▁▁▃▂
time/data_loading,█▄▂▂▁
time/epoch,█▁▁▃▂
train/acc,▁▆▇██
train/loss,█▄▂▁▁
test/acc,0.85742
test/loss,0.50782
time/compute,44.00909
time/data_loading,0.16405


Loading weights:   0%|          | 0/100 [00:00<?, ?it/s]

DistilBertForSequenceClassification LOAD REPORT from: distilbert-base-uncased
Key                     | Status     | 
------------------------+------------+-
vocab_projector.bias    | UNEXPECTED | 
vocab_transform.weight  | UNEXPECTED | 
vocab_layer_norm.bias   | UNEXPECTED | 
vocab_transform.bias    | UNEXPECTED | 
vocab_layer_norm.weight | UNEXPECTED | 
classifier.weight       | MISSING    | 
pre_classifier.bias     | MISSING    | 
pre_classifier.weight   | MISSING    | 
classifier.bias         | MISSING    | 

Notes:
- UNEXPECTED	:can be ignored when loading from different task/architecture; not ok if you expect identical arch.
- MISSING	:those params were newly initialized because missing from the checkpoint. Consider training on your downstream task.


Initial validation loss: 0.6955315750092268, accuracy: 0.50146484375


 20%|██        | 1/5 [00:58<03:55, 58.81s/it]

Epoch 0: train loss: 0.707947114482522, data loading time: 0.17s, compute time: 44.46s


 40%|████      | 2/5 [01:57<02:55, 58.48s/it]

Epoch 1: train loss: 0.7053753212094307, data loading time: 0.15s, compute time: 43.67s


 60%|██████    | 3/5 [02:55<01:56, 58.30s/it]

Epoch 2: train loss: 0.6947689764201641, data loading time: 0.18s, compute time: 43.53s


 80%|████████  | 4/5 [03:53<00:58, 58.29s/it]

Epoch 3: train loss: 0.6932004615664482, data loading time: 0.16s, compute time: 43.79s


100%|██████████| 5/5 [04:51<00:00, 58.30s/it]

Epoch 4: train loss: 0.6958597395569086, data loading time: 0.17s, compute time: 43.57s


test/acc,██▁██
test/loss,█▂▁▁▁
time/compute,█▂▁▃▁
time/data_loading,▅▁█▄▆
time/epoch,█▂▁▃▁
train/acc,█▃▅▄▁
train/loss,█▇▂▁▂
test/acc,0.50342
test/loss,0.6938
time/compute,43.57373
time/data_loading,0.16821


In [1]:
gpu_info = !nvidia-smi
gpu_info = '\n'.join(gpu_info)
if gpu_info.find('failed') >= 0:
  print('Not connected to a GPU')
else:
  print(gpu_info)

Sun Mar  1 01:08:02 2026       
+-----------------------------------------------------------------------------------------+
| NVIDIA-SMI 580.82.07              Driver Version: 580.82.07      CUDA Version: 13.0     |
+-----------------------------------------+------------------------+----------------------+
| GPU  Name                 Persistence-M | Bus-Id          Disp.A | Volatile Uncorr. ECC |
| Fan  Temp   Perf          Pwr:Usage/Cap |           Memory-Usage | GPU-Util  Compute M. |
|                                         |                        |               MIG M. |
|=========================================+========================+======================|
|   0  Tesla T4                       Off |   00000000:00:04.0 Off |                    0 |
| N/A   50C    P8             11W /   70W |       0MiB /  15360MiB |      0%      Default |
|                                         |                        |                  N/A |
+-----------------------------------------+-----

In [ ]:
# C6 - Optimizer experiment
import wandb
from transformers import AutoModelForSequenceClassification

model_name = "distilbert-base-uncased"
max_len = 256
batch_size = 32
num_epochs = 5
learning_rate = 1e-4
num_workers = 2
compile_mode = False
optimizer_list = ["SGD", "Adam"]

for optimizer_name in optimizer_list:
    print(f"\n\n=== Running with optimizer: {optimizer_name} ===")

    wandb.init(
        project="hpml-hw2-llm",
        group="C6",
        name=f"optimizer_{optimizer_name}",
        reinit=True
    )

    wandb.config.update({
        "model_name": model_name,
        "max_len": max_len,
        "batch_size": batch_size,
        "lr": learning_rate,
        "optimizer": optimizer_name,
        "num_workers": num_workers,
        "epochs": num_epochs,
        "compile_mode": compile_mode
    })

    model = AutoModelForSequenceClassification.from_pretrained(
        model_name,
        num_labels=2
    )

    train_dataset, validation_dataset = load_data_and_tokenize()

    train_and_count_paramters(
        model,
        train_dataset,
        validation_dataset,
        num_epochs,
        batch_size,
        learning_rate,
        num_workers,
        optimizer_name,
        compile_mode
    )

    wandb.finish()



=== Running with optimizer: SGD ===


model/trainable_param_elems,▁
model/trainable_param_tensors,▁
model/trainable_param_elems,66955010
model/trainable_param_tensors,104


Loading weights:   0%|          | 0/100 [00:00<?, ?it/s]

DistilBertForSequenceClassification LOAD REPORT from: distilbert-base-uncased
Key                     | Status     | 
------------------------+------------+-
vocab_transform.weight  | UNEXPECTED | 
vocab_transform.bias    | UNEXPECTED | 
vocab_projector.bias    | UNEXPECTED | 
vocab_layer_norm.weight | UNEXPECTED | 
vocab_layer_norm.bias   | UNEXPECTED | 
pre_classifier.bias     | MISSING    | 
classifier.bias         | MISSING    | 
pre_classifier.weight   | MISSING    | 
classifier.weight       | MISSING    | 

Notes:
- UNEXPECTED	:can be ignored when loading from different task/architecture; not ok if you expect identical arch.
- MISSING	:those params were newly initialized because missing from the checkpoint. Consider training on your downstream task.


Trainable parameters: 66,955,010 (across 104 tensors)


  0%|          | 0/1 [00:00<?, ?it/s]

Parameters with gradients (after one backward): 66,955,010 (across 104 tensors)


100%|██████████| 1/1 [00:56<00:00, 56.17s/it]

Epoch 0: train loss: 0.6964747803285718, data loading time: 0.33s, compute time: 41.16s, params w/ grads: 66,955,010 (across 104 tensors)


model/grad_param_elems,▁
model/grad_param_tensors,▁
model/trainable_param_elems,▁
model/trainable_param_tensors,▁
test/acc,▁
test/loss,▁
time/compute,▁
time/data_loading,▁
time/epoch,▁
train/acc,▁
+1,...




=== Running with optimizer: Adam ===


Loading weights:   0%|          | 0/100 [00:00<?, ?it/s]

DistilBertForSequenceClassification LOAD REPORT from: distilbert-base-uncased
Key                     | Status     | 
------------------------+------------+-
vocab_transform.weight  | UNEXPECTED | 
vocab_transform.bias    | UNEXPECTED | 
vocab_projector.bias    | UNEXPECTED | 
vocab_layer_norm.weight | UNEXPECTED | 
vocab_layer_norm.bias   | UNEXPECTED | 
pre_classifier.bias     | MISSING    | 
classifier.bias         | MISSING    | 
pre_classifier.weight   | MISSING    | 
classifier.weight       | MISSING    | 

Notes:
- UNEXPECTED	:can be ignored when loading from different task/architecture; not ok if you expect identical arch.
- MISSING	:those params were newly initialized because missing from the checkpoint. Consider training on your downstream task.


Trainable parameters: 66,955,010 (across 104 tensors)


  0%|          | 0/1 [00:00<?, ?it/s]

Parameters with gradients (after one backward): 66,955,010 (across 104 tensors)


100%|██████████| 1/1 [00:57<00:00, 57.63s/it]

Epoch 0: train loss: 0.44779287721030414, data loading time: 0.31s, compute time: 43.08s, params w/ grads: 66,955,010 (across 104 tensors)


model/grad_param_elems,▁
model/grad_param_tensors,▁
model/trainable_param_elems,▁
model/trainable_param_tensors,▁
test/acc,▁
test/loss,▁
time/compute,▁
time/data_loading,▁
time/epoch,▁
train/acc,▁
+1,...


In [9]:
# C7 - Compile mode experiment
import wandb
from transformers import AutoModelForSequenceClassification

model_name = "distilbert-base-uncased"
max_len = 256
batch_size = 32
num_epochs = 10
learning_rate = 1e-4
num_workers = 0
optimizer_name = "AdamW"

compile_mode_list = [False, True]

for compile_mode in compile_mode_list:

    wandb.init(
        project="hpml-hw2-llm",
        group="C7",
        name=f"compile_{compile_mode}",
        reinit=True
    )

    wandb.config.update({
        "model_name": model_name,
        "max_len": max_len,
        "batch_size": batch_size,
        "lr": learning_rate,
        "optimizer": optimizer_name,
        "num_workers": num_workers,
        "epochs": num_epochs,
        "compile_mode": compile_mode
    })

    model = AutoModelForSequenceClassification.from_pretrained(
        model_name,
        num_labels=2
    )

    # 关键：根据 compile_mode 决定是否 torch.compile
    if compile_mode:
        import torch
        model = torch.compile(model)

    train_dataset, validation_dataset = load_data_and_tokenize()

    train(
        model,
        train_dataset,
        validation_dataset,
        num_epochs,
        batch_size,
        learning_rate,
        num_workers,
        optimizer_name
    )

    wandb.finish()

Loading weights:   0%|          | 0/100 [00:00<?, ?it/s]

DistilBertForSequenceClassification LOAD REPORT from: distilbert-base-uncased
Key                     | Status     | 
------------------------+------------+-
vocab_transform.weight  | UNEXPECTED | 
vocab_transform.bias    | UNEXPECTED | 
vocab_projector.bias    | UNEXPECTED | 
vocab_layer_norm.weight | UNEXPECTED | 
vocab_layer_norm.bias   | UNEXPECTED | 
pre_classifier.bias     | MISSING    | 
classifier.bias         | MISSING    | 
pre_classifier.weight   | MISSING    | 
classifier.weight       | MISSING    | 

Notes:
- UNEXPECTED	:can be ignored when loading from different task/architecture; not ok if you expect identical arch.
- MISSING	:those params were newly initialized because missing from the checkpoint. Consider training on your downstream task.


Initial validation loss: 0.690304558724165, accuracy: 0.541015625


 10%|█         | 1/10 [00:56<08:29, 56.61s/it]

Epoch 0: train loss: 0.6031508673913777, data loading time: 0.18s, compute time: 42.06s


 20%|██        | 2/10 [01:53<07:34, 56.86s/it]

Epoch 1: train loss: 0.33922610303852707, data loading time: 0.18s, compute time: 42.95s


 30%|███       | 3/10 [02:50<06:38, 56.97s/it]

Epoch 2: train loss: 0.19463405496207997, data loading time: 0.19s, compute time: 42.84s


 40%|████      | 4/10 [03:47<05:42, 57.04s/it]

Epoch 3: train loss: 0.11339166067773476, data loading time: 0.19s, compute time: 42.86s


 50%|█████     | 5/10 [04:44<04:45, 57.02s/it]

Epoch 4: train loss: 0.05862562861148035, data loading time: 0.19s, compute time: 42.72s


 60%|██████    | 6/10 [05:42<03:48, 57.10s/it]

Epoch 5: train loss: 0.033435151337471325, data loading time: 0.18s, compute time: 42.93s


 70%|███████   | 7/10 [06:39<02:51, 57.12s/it]

Epoch 6: train loss: 0.03082800540869357, data loading time: 0.18s, compute time: 42.89s


 80%|████████  | 8/10 [07:36<01:54, 57.12s/it]

Epoch 7: train loss: 0.029153118033718783, data loading time: 0.18s, compute time: 42.93s


 90%|█████████ | 9/10 [08:33<00:57, 57.10s/it]

Epoch 8: train loss: 0.04383226003847085, data loading time: 0.19s, compute time: 42.76s


100%|██████████| 10/10 [09:30<00:00, 57.05s/it]

Epoch 9: train loss: 0.022285436194579233, data loading time: 0.19s, compute time: 42.86s


test/acc,▇▁▂█▇███▆▄
test/loss,▁▁▃▃▄▄▄▇▅█
time/compute,▁█▇▇▆███▇▇
time/data_loading,▄▄█▅▇▁▁▁▅▅
time/epoch,▁█▇▇▆███▇▇
train/acc,▁▅▇▇██████
train/loss,█▅▃▂▁▁▁▁▁▁
test/acc,0.82227
test/loss,0.83392
time/compute,42.85829
time/data_loading,0.18575


Loading weights:   0%|          | 0/100 [00:00<?, ?it/s]

DistilBertForSequenceClassification LOAD REPORT from: distilbert-base-uncased
Key                     | Status     | 
------------------------+------------+-
vocab_transform.weight  | UNEXPECTED | 
vocab_transform.bias    | UNEXPECTED | 
vocab_projector.bias    | UNEXPECTED | 
vocab_layer_norm.weight | UNEXPECTED | 
vocab_layer_norm.bias   | UNEXPECTED | 
pre_classifier.bias     | MISSING    | 
classifier.bias         | MISSING    | 
pre_classifier.weight   | MISSING    | 
classifier.weight       | MISSING    | 

Notes:
- UNEXPECTED	:can be ignored when loading from different task/architecture; not ok if you expect identical arch.
- MISSING	:those params were newly initialized because missing from the checkpoint. Consider training on your downstream task.
W0301 21:18:40.994000 420 torch/_inductor/utils.py:1679] [0/0] Not enough SMs to use max_autotune_gemm mode
/usr/local/lib/python3.12/dist-packages/torch/_inductor/lowering.py:7627: UserWarning: 
Online softmax is disabled on the fly 

Initial validation loss: 0.6886299680918455, accuracy: 0.50439453125


  0%|          | 0/10 [00:00<?, ?it/s]/usr/local/lib/python3.12/dist-packages/torch/_inductor/lowering.py:7627: UserWarning: 
Online softmax is disabled on the fly since Inductor decides to
split the reduction. Cut an issue to PyTorch if this is an
important use case and you want to speed it up with online
softmax.

  warnings.warn(
 10%|█         | 1/10 [01:19<11:59, 79.90s/it]

Epoch 0: train loss: 0.40546627109870315, data loading time: 0.17s, compute time: 65.99s


 20%|██        | 2/10 [02:16<08:47, 65.96s/it]

Epoch 1: train loss: 0.18313575600041077, data loading time: 0.18s, compute time: 42.14s


 30%|███       | 3/10 [03:12<07:10, 61.53s/it]

Epoch 2: train loss: 0.11764192629925674, data loading time: 0.18s, compute time: 42.32s


 40%|████      | 4/10 [04:08<05:56, 59.46s/it]

Epoch 3: train loss: 0.06883976853714557, data loading time: 0.17s, compute time: 42.29s


 50%|█████     | 5/10 [05:04<04:51, 58.29s/it]

Epoch 4: train loss: 0.03308609714440536, data loading time: 0.18s, compute time: 42.25s


 60%|██████    | 6/10 [06:01<03:50, 57.66s/it]

Epoch 5: train loss: 0.03188803844568611, data loading time: 0.18s, compute time: 42.44s


 70%|███████   | 7/10 [06:57<02:51, 57.21s/it]

Epoch 6: train loss: 0.034689860192884225, data loading time: 0.18s, compute time: 42.31s


 80%|████████  | 8/10 [07:53<01:53, 56.93s/it]

Epoch 7: train loss: 0.0482980995602702, data loading time: 0.17s, compute time: 42.42s


 90%|█████████ | 9/10 [08:50<00:56, 56.73s/it]

Epoch 8: train loss: 0.043821174102049554, data loading time: 0.17s, compute time: 42.24s


100%|██████████| 10/10 [09:46<00:00, 58.65s/it]

Epoch 9: train loss: 0.028844621952885063, data loading time: 0.18s, compute time: 42.38s


test/acc,▆█▇▄▁▇▄▁▇▇
test/loss,▁▁▂▄▆▄▆█▃▂
time/compute,█▁▁▁▁▁▁▁▁▁
time/data_loading,▂▃▄▂▃▃▄▁▁█
time/epoch,█▁▁▁▁▁▁▁▁▁
train/acc,▁▆▇▇██████
train/loss,█▄▃▂▁▁▁▁▁▁
test/acc,0.86572
test/loss,0.39936
time/compute,42.37622
time/data_loading,0.18445


# HPML HW2 REPORT

---
## Environment Summary
- GPU: Tesla T4, 15360 MiB, 580.82.07
- OS: Linux 6.6.113+ (#1 SMP Mon Feb  2 12:27:57 UTC 2026)
- Python: 3.12.12 (main, Oct 10 2025, 08:52:57) [GCC 11.4.0]
- PyTorch: 2.10.0+cu128 (CUDA: 12.8, cuDNN: 91002)
- Transformers: 5.0.0
- Datasets: 4.0.0
- W&B: 0.25.0

--- 
## WanDB 
Project Link: https://wandb.ai/desolate0602-columbia-university/hpml-hw2-llm/

<div align="center">
  <img src="https://raw.githubusercontent.com/jasonlee-1024/image-host/main/hw2/wandb_overview.png" width="80%" />
</div>

---

## C1 Fine-tuning a Small LLM

### Training Hyperparameters
- Batch Size: 32
- Number of Epochs: 5
- Learning Rate: 1e-4
- Optimizer: AdamW
- Compile Mode: False

### Results

<div align="center">
  <img src="https://raw.githubusercontent.com/jasonlee-1024/image-host/main/hw2/C1-Training_Loss_Curve.png" width="45%" />
  <img src="https://raw.githubusercontent.com/jasonlee-1024/image-host/main/hw2/C1-Training_Accuracy_Curve.png" width="45%" />
</div>

<p align="center">
Loss Curve (left) and Accuracy Curve (right)
</p>

---
## C2 Baseline Timing

### Training Hyperparameters
- Batch Size: 32
- Number of Epochs: 5
- Learning Rate: 1e-4
- Optimizer: AdamW
- Compile Mode: False

### Results

<div align="center">
  <img src="https://raw.githubusercontent.com/jasonlee-1024/image-host/main/hw2/C2_time_epoch.png" width="32%" />
  <img src="https://raw.githubusercontent.com/jasonlee-1024/image-host/main/hw2/C2_time_data.png" width="32%" />
  <img src="https://raw.githubusercontent.com/jasonlee-1024/image-host/main/hw2/C2_time_compute.png" width="32%" />
</div>

<div align="center">
<b>Figure</b> Epoch time (left), data loading time (middle), and compute time (right) for C2.
</div>

---
## C3: Effect of DataLoader Workers

### Experiment Configuration

| Category | Parameter | Value |
|----------|-----------|-------|
| Model | Model Name | distilbert-base-uncased |
| Model | Max Length | 256 |
| Training | Batch Size | 32 |
| Training | Epochs | 5 |
| Training | Learning Rate | 1e-4 |
| Training | Optimizer | AdamW |
| System | Compile Mode | False |
| System | DataLoader Workers | 0, 2, 4, 8 |
| Dataset | Train Data Size | 2048 |
| Dataset | Test Data Size | 2048 |

### Performance Results

<div align="center">
  <img src="https://raw.githubusercontent.com/jasonlee-1024/image-host/main/hw2/C3_time_epoch.png" width="45%" />
  <img src="https://raw.githubusercontent.com/jasonlee-1024/image-host/main/hw2/C3_time_data.png" width="45%" />
</div>

<div align="center">
<b>Figure:</b> Epoch time (left) and data loading time (right) under different numbers of DataLoader workers.
</div>

### Summary

1. Since the dataset size is relatively small and samples are already pre-tokenized, increasing the number of DataLoader workers introduces additional multiprocessing overhead (process creation, inter-process communication), which leads to longer data loading time instead of speedup.

2. As the overall training time is dominated by GPU computation rather than data loading, changes in the number of workers do not significantly affect the total epoch time.

---

## C4: Pytorch Profiler

### Experiment Configuration

| Parameter       | Value                     |
|-----------------|---------------------------|
| Model Name      | distilbert-base-uncased   |
| Max Length      | 256                       |
| Batch Size      | 32                        |
| Number of Epochs| 1                         |
| Learning Rate   | 1e-4                      |
| Optimizer       | AdamW                     |
| Num Workers     | 0,1                       |
| Compile Mode    | False                     |
| Train Data Size | 2048                      |
| Train Data Size | 2048                      |

### TensorBoard Report (0 Worker)

<div align="center">
  <img src="https://raw.githubusercontent.com/jasonlee-1024/image-host/main/hw2/c4_summary_worker_0.png" width="80%" />
</div>

<div align="center">
<b>Figure:</b> TensorBoard report, num_works = 0
</div>

### TensorBoard Report (1 Worker)

<div align="center">
  <img src="https://raw.githubusercontent.com/jasonlee-1024/image-host/main/hw2/c4_summary_worker_1.png" width="80%" />
</div>

<div align="center">
<b>Figure:</b> TensorBoard report, num_works = 1
</div>

### Comparison Table
<div align="center">
  <img src="https://raw.githubusercontent.com/jasonlee-1024/image-host/main/hw2/c4_time_worker_0.png" width="45%" />
  <img src="https://raw.githubusercontent.com/jasonlee-1024/image-host/main/hw2/c4_time_worker_1.png" width="45%" />
</div>

<div align="center">
<b>Figure: </b>Comparison between 0 worker (left) and 1 worker (right)
</div>

### Performance Analysis

1. Since the dataset is relatively small and has already been tokenized, data loading introduces negligible overhead and has minimal impact on overall training time.

2. Compared to data loading, the vast majority of execution time is spent on GPU computation, accounting for more than 99% of the total runtime.

3. The primary bottleneck lies in computational throughput rather than in data input or preprocessing.

---
## C5: Effect of Batch Size and Learning Rate

### Experiment Configuration

| Category | Parameter | Value |
|----------|-----------|-------|
| Model | Model Name | distilbert-base-uncased |
| Model | Max Length | 256 |
| Training | Epochs | 5 |
| Training | Optimizer | AdamW |
| System | Compile Mode | False |
| System | DataLoader Workers | 2 |
| Dataset | Train Data Size | 2048 |
| Dataset | Test Data Size | 2048 |

### Hyperparameter Sweep

- Batch Size: 16, 32, 64  
- Learning Rate: 5e-5, 1e-4, 5e-4  


### Performance Results

#### Time Analysis

<div align="center">
  <img src="https://raw.githubusercontent.com/jasonlee-1024/image-host/main/hw2/C5_time.png" width="45%" />
</div>

<div align="center">
<b>Figure 1:</b> Training time under different batch size and learning rate settings.
</div>


#### Loss Curves

<div align="center">
  <img src="https://raw.githubusercontent.com/jasonlee-1024/image-host/main/hw2/C5_train_loss.png" width="45%" />
  <img src="https://raw.githubusercontent.com/jasonlee-1024/image-host/main/hw2/C5_test_loss.png" width="45%" />
</div>

<div align="center">
<b>Figure 2:</b> Training loss (left) and test loss (right) under different hyperparameter settings.
</div>


#### Training Accuracy

<div align="center">
  <img src="https://raw.githubusercontent.com/jasonlee-1024/image-host/main/hw2/C5_train_accuracy.png" width="45%" />
</div>

<div align="center">
<b>Figure 3:</b> Training accuracy under different hyperparameter settings.
</div>

### Summary and Analysis

From the results, we observe that increasing the batch size and learning rate generally improves training speed, as larger batches increase computational throughput and higher learning rates accelerate parameter updates. 

However, this speed improvement comes at the cost of model performance and stability. Larger batch sizes and higher learning rates lead to noisier optimization dynamics, resulting in reduced accuracy and less stable convergence behavior, as reflected in the training and test loss curves.

---

## C6: Optimizer Comparison

### Experiment Configuration

| Category | Parameter | Value |
|----------|-----------|-------|
| Model | Model Name | distilbert-base-uncased |
| Model | Max Length | 256 |
| Training | Epochs | 5 |
| Training | Batch Size | 32 |
| Training | Learning Rate | 1e-4 |
| System | Compile Mode | False |
| System | DataLoader Workers | 2 |
| Dataset | Train Data Size | 2048 |
| Dataset | Test Data Size | 2048 |

### Hyperparameter

- Optimizer: Adam, SGD, AdamW


### Performance Results

#### Time Analysis

<div align="center">
  <img src="https://raw.githubusercontent.com/jasonlee-1024/image-host/main/hw2/C6_time_epoch.png" width="45%" />
</div>

<div align="center">
<b>Figure 1:</b> Training time under different Optimizer.
</div>


#### Loss Curves

<div align="center">
  <img src="https://raw.githubusercontent.com/jasonlee-1024/image-host/main/hw2/C6_train_loss.png" width="45%" />
  <img src="https://raw.githubusercontent.com/jasonlee-1024/image-host/main/hw2/C6_test_loss.png" width="45%" />
</div>

<div align="center">
<b>Figure 2:</b> Training loss (left) and test loss (right) under different hyperparameter settings.
</div>


#### Training Accuracy

<div align="center">
  <img src="https://raw.githubusercontent.com/jasonlee-1024/image-host/main/hw2/C6_train_accuracy.png" width="45%" />
  <img src="https://raw.githubusercontent.com/jasonlee-1024/image-host/main/hw2/C6_test_accuracy.png" width="45%" />
</div>

<div align="center">
<b>Figure 3:</b> Training accuracy under different hyperparameter settings.
</div>

---

## C7: Compiler Mode

### Experiment Configuration

| Category | Parameter | Value |
|----------|-----------|-------|
| Model | Model Name | distilbert-base-uncased |
| Model | Max Length | 256 |
| Training | Epochs | 10 |
| Training | Batch Size | 32 |
| Training | Learning Rate | 1e-4 |
| Training | Optimizer | AdamW |
| System | Compile Mode | False |
| System | DataLoader Workers | 2 |
| Dataset | Train Data Size | 2048 |
| Dataset | Test Data Size | 2048 |

### Hyperparameter

- Compile Mode: False, True

### Results

| Mode                | First Epoch Time (s) | Avg. Time Epochs 6–10 (s) |
|---------------------|----------------------|----------------------------|
| Eager               |       42.2           |       42.8                 |
| Compile (Inductor)  |       66.1           |       42.4                 |

---


# Short Answer Question

#### Q1. What is the input dimension of DistilBERT’s embedding layer?

The input dimension of DistilBERT’s embedding layer is the vocabulary size.

#### Q2. What is the output dimension of the classifier head for IMDB?
It is a binary classification task, so the classifier head outputs logits of shape (batch_size, 2).

#### Q3. Report the number of trainable parameters and parameters with gradients after a backward pass. Include the code you used to compute both.

Trainable parameters: 66,955,010 (across 104 tensors)
Parameters with gradients (after one backward): 66,955,010 (across 104 tensors)

```python
def count_trainable_params(model: torch.nn.Module):
    trainable_tensors = 0
    trainable_elems = 0
    for p in model.parameters():
        if p.requires_grad:
            trainable_tensors += 1
            trainable_elems += p.numel()
    return trainable_tensors, trainable_elems


def count_params_with_grads(model: torch.nn.Module):
    grad_tensors = 0
    grad_elems = 0
    for p in model.parameters():
        if p.grad is not None:
            grad_tensors += 1
            grad_elems += p.grad.numel()
    return grad_tensors, grad_elems
```
#### Q4. How (if at all) does parameter count change when switching from SGD to Adam?

No change

### Q5. Why is the first epoch slower with torch.compile, and why are later epochs typically faster?
The first epoch is slower with torch.compile because it must capture, optimize, and compile the computation graph during the initial execution, while later epochs reuse the optimized graph and therefore run faster with reduced overhead and improved kernel efficiency.